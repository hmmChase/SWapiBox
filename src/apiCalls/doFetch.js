export const doFetch = async url => {
  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`${response.status}`);
    }
    return await response.json();
  } catch (error) {
    throw new Error(`Network request failed. (error: ${error.message})`);
  }
};
