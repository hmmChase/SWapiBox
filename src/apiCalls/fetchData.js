import { fetchPeople } from './peopleData';
import { fetchPlanets } from './planetData';
import { fetchVehicles } from './vehicleData';

export async function fetchCategoryData(category) {
  switch (category) {
  case 'people':
    return fetchPeople(category);
  case 'planets':
    return fetchPlanets(category);
  case 'vehicles':
    return fetchVehicles(category);
  default:
    break;
  }
}

export async function makeFetch(URL) {
  try {
    const response = await fetch(URL);
    return await response.json();
  } catch (error) {
    throw new Error(error);
  }
}