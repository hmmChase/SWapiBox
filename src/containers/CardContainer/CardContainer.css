.Card-container {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
}

.pretext {
  display: flex;
  justify-content: center;
  align-items: center;
  overflow: hidden;
  height: 80vh;
  font-size: 2em;
}
