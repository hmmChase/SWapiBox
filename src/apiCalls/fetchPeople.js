import { doFetch } from './doFetch';

export const fetchPeople = async () => {
  const root = 'http://swapi.com/api';
  const peopleData = await doFetch(`${root}/people`);
  return peopleData;
};
