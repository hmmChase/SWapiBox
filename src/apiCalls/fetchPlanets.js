import { doFetch } from './doFetch';

export const fetchPlanets = async () => {
  try {
    const root = 'https://swapi.co/api';
    const planetsData = await doFetch(`${root}/planets`);
    const parsedPlanetsData = planetsData.results.map(async planet => {
      const residentsData = planet.residents.map(async residentURL => {
        return await fetchResidents(residentURL);
      });
      const resolveResidents = await Promise.all(residentsData);
      return {
        name: planet.name,
        terrain: planet.terrain,
        population: planet.population,
        climate: planet.climate,
        residents: resolveResidents.length
          ? resolveResidents.join(', ')
          : 'unknown'
      };
    });
    return await Promise.all(parsedPlanetsData);
  } catch (error) {
    throw new Error(error.message);
  }
};

export const fetchResidents = async residentURL => {
  try {
    const residentData = await doFetch(residentURL);
    return residentData.name;
  } catch (error) {
    throw new Error(error.message);
  }
};
