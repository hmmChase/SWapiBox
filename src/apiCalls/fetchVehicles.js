import { doFetch } from './doFetch';

export const fetchVehicles = async () => {
  const root = 'https://swapi.co/api';
  const vehiclesData = await doFetch(`${root}/vehicles`);
  return vehiclesData.results.map(vehicle => {
    return {
      name: vehicle.name,
      model: vehicle.model,
      class: vehicle.vehicle_class,
      passengers: vehicle.passengers
    };
  });
}
