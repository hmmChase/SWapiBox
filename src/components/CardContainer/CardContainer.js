import React, { Component } from 'react';
import Card from '../Card/Card';
import './CardContainer.css';
import { fetchPeople } from '../../apiCalls/fetchPeople';
import { fetchPlanets } from '../../apiCalls/fetchPlanets';
import { fetchVehicles } from '../../apiCalls/fetchVehicles';

class CardContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      category: this.props.match.path.slice(1),
      people: [],
      planets: [],
      vehicles: [],
      favorites: []
    };
  }

  async componentDidMount() {
    await this.getCategoryData(this.state.category);
  }

  getCategoryData = async category => {
    if (this.state[category].length === 0) {
      switch (category) {
        case 'people':
          const people = await fetchPeople();
          this.setState({ people });
          break;
        case 'planets':
          const planets = await fetchPlanets();
          this.setState({ planets });
          break;
        case 'vehicles':
          const vehicles = await fetchVehicles();
          this.setState({ vehicles });
          break;
        default:
          break;
      }
    }
  };

  renderCards = () => {
    if (this.state[this.state.category].length > 0) {
      return this.state[this.state.category].map((dataObj, index) => {
        return <Card key={'card' + index} dataObj={dataObj} />;
      });
    }
    return <p>...loading</p>;
  };

  render() {
    return <section className="Card-container">{this.renderCards()}</section>;
  }
}

export default CardContainer;
