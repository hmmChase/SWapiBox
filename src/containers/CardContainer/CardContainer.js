import React, { Component } from 'react';
import Card from '../../components/Card/Card';
import PropTypes from 'prop-types';
import './CardContainer.css';
import { fetchPeople } from '../../apiCalls/fetchPeople';
import { fetchPlanets } from '../../apiCalls/fetchPlanets';
import { fetchVehicles } from '../../apiCalls/fetchVehicles';

class CardContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      category: this.props.match.path.slice(1),
      people: [],
      planets: [],
      vehicles: [],
      favorites: []
    };
  }

  async componentDidMount() {
    await this.getCategoryData(this.state.category);
    this.retrieveFavorites();
  }

  getCategoryData = async category => {
    if (this.state[category].length === 0) {
      switch (category) {
        case 'people':
          this.setState({ people: await fetchPeople() });
          break;
        case 'planets':
          this.setState({ planets: await fetchPlanets() });
          break;
        case 'vehicles':
          this.setState({ vehicles: await fetchVehicles() });
          break;
        default:
          break;
      }
    }
  };

  retrieveFavorites() {
    if (localStorage.favorites) {
      const localFavorites = JSON.parse(localStorage.getItem('favorites'));
      this.setState({
        favorites: localFavorites
      });
    }
  }

  toggleFavorite = favoriteObj => {
    const alreadyFavorite = this.state.favorites.some(
      favorite => favorite.name === favoriteObj.name
    );
    if (alreadyFavorite) {
      this.removeFavorite(favoriteObj);
    } else {
      this.addFavorite(favoriteObj);
    }
    this.props.updateFavAmt();
  };

  removeFavorite(favoriteObj) {
    const deletedfavoriteObj = this.state.favorites.filter(
      favorite => favorite.name !== favoriteObj.name
    );
    localStorage.setItem('favorites', JSON.stringify(deletedfavoriteObj));
    this.setState({ favorites: deletedfavoriteObj });
  }

  addFavorite(favoriteObj) {
    localStorage.setItem(
      'favorites',
      JSON.stringify([...this.state.favorites, favoriteObj])
    );
    this.setState({
      favorites: [...this.state.favorites, favoriteObj]
    });
  }

  createCards = () => {
    return this.state[this.state.category].map((dataObj, index) => {
      const favoriteCardNames = this.state.favorites.map(
        favorite => favorite.name
      );
      const favoriteBool = favoriteCardNames.includes(dataObj.name);
      return (
        <Card
          key={'card' + index}
          dataObj={dataObj}
          toggleFavorite={this.toggleFavorite}
          favoriteBool={favoriteBool}
        />
      );
    });
  };

  render() {
    return (
      <section className="Card-container">
        {this.state[this.state.category].length > 0 ? (
          this.createCards()
        ) : this.state.category === 'favorites' ? (
          <p className="pretext">Favorites, you need</p>
        ) : (
          <p className="pretext">...loading</p>
        )}
      </section>
    );
  }
}

CardContainer.propTypes = {
  match: PropTypes.object.isRequired,
  updateFavAmt: PropTypes.func.isRequired
};

export default CardContainer;
