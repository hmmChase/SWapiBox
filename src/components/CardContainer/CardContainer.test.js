import React from 'react';
import { shallow } from 'enzyme';
import CardContainer from './CardContainer';
import { fetchPeople } from '../../apiCalls/fetchPeople';
jest.mock('../../apiCalls/fetchPeople.js');
import { fetchPlanets } from '../../apiCalls/fetchPlanets';
jest.mock('../../apiCalls/fetchPlanets.js');
import { fetchVehicles } from '../../apiCalls/fetchVehicles';
jest.mock('../../apiCalls/fetchVehicles.js');

describe('CardContainer', () => {
  let mockProps;
  let cardContainer;
  beforeEach(() => {
    jest.resetAllMocks();
    mockProps = {
      match: { path: '/vehicles' }
    };
    cardContainer = shallow(<CardContainer {...mockProps} />, {
      disableLifecycleMethods: true
    });
  });

  it('matches snapshot', () => {
    expect(cardContainer).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    it('calls getCategoryData with correct arg', async () => {
      const getCategoryData = (cardContainer.instance().getCategoryData = jest.fn());

      // why doesn't this work??
      // cardContainer = shallow(<CardContainer {...mockProps} />);
      await cardContainer.instance().componentDidMount();

      await expect(getCategoryData).toHaveBeenCalledTimes(1);
      await expect(getCategoryData).toHaveBeenCalledWith('vehicles');
    });
  });
  describe('getCategoryData', () => {
    it('if people data not loaded, calls fetchPeople based on people category', async () => {
      await cardContainer.instance().getCategoryData('people');

      await expect(fetchPeople).toHaveBeenCalledTimes(1);
    });

    it('if planets data not loaded, calls fetchPlanets based on planets category', async () => {
      await cardContainer.instance().getCategoryData('planets');

      await expect(fetchPlanets).toHaveBeenCalledTimes(1);
    });

    it('if vehicles data not loaded, calls fetchVehicles based on vehicles category', async () => {
      console.log(cardContainer.state());

      await cardContainer.instance().getCategoryData('vehicles');

      await expect(fetchVehicles).toHaveBeenCalledTimes(1);
    });

    // it('does nothing if category is invalid', async () => {
    //   const category = 'lkjasdf';
    //   const invalidCategory = await getCategoryData(category);

    //   expect(invalidCategory).toEqual(null);
    // });

    describe('renderCards', () => {
      it('returns a Card component for every cardObj', () => {
        const mockVehicleData = [
          {
            name: 'Sand Crawler',
            model: 'Digger Crawler',
            class: 'wheeled',
            passengers: '30'
          },
          {
            name: 'T-16 skyhopper',
            model: 'T-16 skyhopper',
            class: 'repulsorcraft',
            passengers: '1'
          },
          {
            name: 'X-34 landspeeder',
            model: 'X-34 landspeeder',
            class: 'repulsorcraft',
            passengers: '1'
          },
          {
            name: 'TIE/LN starfighter',
            model: 'Twin Ion Engine/Ln Starfighter',
            class: 'starfighter',
            passengers: '0'
          },
          {
            name: 'Snowspeeder',
            model: 't-47 airspeeder',
            class: 'airspeeder',
            passengers: '0'
          },
          {
            name: 'TIE bomber',
            model: 'TIE/sa bomber',
            class: 'space/planetary bomber',
            passengers: '0'
          },
          {
            name: 'AT-AT',
            model: 'All Terrain Armored Transport',
            class: 'assault walker',
            passengers: '40'
          },
          {
            name: 'AT-ST',
            model: 'All Terrain Scout Transport',
            class: 'walker',
            passengers: '0'
          },
          {
            name: 'Storm IV Twin-Pod cloud car',
            model: 'Storm IV Twin-Pod',
            class: 'repulsorcraft',
            passengers: '0'
          },
          {
            name: 'Sail barge',
            model: 'Modified Luxury Sail Barge',
            class: 'sail barge',
            passengers: '500'
          }
        ];
        cardContainer.setState({ vehicles: mockVehicleData });
        const renderCards = cardContainer.instance().renderCards();

        expect(renderCards).toHaveLength(10);
      });
    });
  });
});
