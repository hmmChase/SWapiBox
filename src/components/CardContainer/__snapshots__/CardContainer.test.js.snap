// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContainer matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CardContainer
    match={
        Object {
            "path": "/vehicles",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        ...loading
</p>,
      "className": "Card-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "...loading",
      },
      "ref": null,
      "rendered": "...loading",
      "type": "p",
    },
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          ...loading
</p>,
        "className": "Card-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "...loading",
        },
        "ref": null,
        "rendered": "...loading",
        "type": "p",
      },
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
